### DISCUSSION TITLE
What makes up the 'current day' ?

### DISCUSSION THREAD

**Discussion by d1964b** [BottlecapDave/HomeAssistant-OctopusEnergy#314] [Q&A] [answered]
I'm overthinking this I'm sure but what is the "current_day_average_rate" actually calculating ? The value changes every 30 mins or so , so  it can't just be the average rate for todays date as that would only change once every 24 hrs. So is it the average rate for the last 24 hrs or for as much of the next 24 ( at least where those prices are known ).   Is the concept of what constitutes the 'current_day' the same across all 'current_day' sensors ?

**Comment by BottlecapDave** [comment, marked as answer]
Hi there. The `current_day_average_rate` located on `current_rate` is indeed the average price for today (midnight to midnight) and _should_ only change once a day. If you're on agile it _could_ change multiple times during the day as new rates come in. What sort of changes are you seeing? what tariff are you on?

**Comment by d1964b** [comment]
I am on agile , but I only get new prices once a day about 1600, though I only get rates up to tomorrow 22:30 so its a bit weird. Any I'll keep an eye on it. The current price tracks OK , which to be honest is the only rate I really rely on. The average rate is really only for interest only.

**Reply by deltamelter** [reply]
I take average from 22:30 and the 24 hours backwards to get what I see as a better average for the coming day.  After midnight/(1am) we start a new day and the entity truncates from midnight so the average is of less hours and the average changes, but it's just a rough guide so works for me.
![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/27061628/48845e19-a913-4363-8b1c-11f8b5123452)

**Reply by BottlecapDave** [reply]
I went for current day as this is easy to understand and the sensor has to translate across all different tariffs. I'm not wanting special cases for when the meaning of sensors change, as that just causes confusion and more issues raised. If I went for last 24 hours, then this would change constantly. I can also only work with the data that OE provides.

However, I would only expect this data to be _slightly_ out throughout the day. For example, https://api.octopus.energy/v1/products/AGILE-FLEX-BB-23-02-08/electricity-tariffs/E-1R-AGILE-FLEX-BB-23-02-08-C/standard-unit-rates/?period_from=2023-07-18T23:00:00Z&period_to=2023-07-19T23:00:00Z at the current time returns the following data

```
{"count":46,"next":null,"previous":null,"results":[{"value_exc_vat":15.52,"value_inc_vat":16.296,"valid_from":"2023-07-19T21:30:00Z","valid_to":"2023-07-19T22:00:00Z","payment_method":null},{"value_exc_vat":19.0,"value_inc_vat":19.95,"valid_from":"2023-07-19T21:00:00Z","valid_to":"2023-07-19T21:30:00Z","payment_method":null},{"value_exc_vat":20.82,"value_inc_vat":21.861,"valid_from":"2023-07-19T20:30:00Z","valid_to":"2023-07-19T21:00:00Z","payment_method":null},{"value_exc_vat":20.84,"value_inc_vat":21.882,"valid_from":"2023-07-19T20:00:00Z","valid_to":"2023-07-19T20:30:00Z","payment_method":null},{"value_exc_vat":19.5,"value_inc_vat":20.475,"valid_from":"2023-07-19T19:30:00Z","valid_to":"2023-07-19T20:00:00Z","payment_method":null},{"value_exc_vat":21.24,"value_inc_vat":22.302,"valid_from":"2023-07-19T19:00:00Z","valid_to":"2023-07-19T19:30:00Z","payment_method":null},{"value_exc_vat":19.78,"value_inc_vat":20.769,"valid_from":"2023-07-19T18:30:00Z","valid_to":"2023-07-19T19:00:00Z","payment_method":null},{"value_exc_vat":21.62,"value_inc_vat":22.701,"valid_from":"2023-07-19T18:00:00Z","valid_to":"2023-07-19T18:30:00Z","payment_method":null},{"value_exc_vat":31.4,"value_inc_vat":32.97,"valid_from":"2023-07-19T17:30:00Z","valid_to":"2023-07-19T18:00:00Z","payment_method":null},{"value_exc_vat":31.0,"value_inc_vat":32.55,"valid_from":"2023-07-19T17:00:00Z","valid_to":"2023-07-19T17:30:00Z","payment_method":null},{"value_exc_vat":29.7,"value_inc_vat":31.185,"valid_from":"2023-07-19T16:30:00Z","valid_to":"2023-07-19T17:00:00Z","payment_method":null},{"value_exc_vat":27.2,"value_inc_vat":28.56,"valid_from":"2023-07-19T16:00:00Z","valid_to":"2023-07-19T16:30:00Z","payment_method":null},{"value_exc_vat":27.18,"value_inc_vat":28.539,"valid_from":"2023-07-19T15:30:00Z","valid_to":"2023-07-19T16:00:00Z","payment_method":null},{"value_exc_vat":26.0,"value_inc_vat":27.3,"valid_from":"2023-07-19T15:00:00Z","valid_to":"2023-07-19T15:30:00Z","payment_method":null},{"value_exc_vat":12.56,"value_inc_vat":13.188,"valid_from":"2023-07-19T14:30:00Z","valid_to":"2023-07-19T15:00:00Z","payment_method":null},{"value_exc_vat":11.76,"value_inc_vat":12.348,"valid_from":"2023-07-19T14:00:00Z","valid_to":"2023-07-19T14:30:00Z","payment_method":null},{"value_exc_vat":11.76,"value_inc_vat":12.348,"valid_from":"2023-07-19T13:30:00Z","valid_to":"2023-07-19T14:00:00Z","payment_method":null},{"value_exc_vat":12.66,"value_inc_vat":13.293,"valid_from":"2023-07-19T13:00:00Z","valid_to":"2023-07-19T13:30:00Z","payment_method":null},{"value_exc_vat":11.5,"value_inc_vat":12.075,"valid_from":"2023-07-19T12:30:00Z","valid_to":"2023-07-19T13:00:00Z","payment_method":null},{"value_exc_vat":12.0,"value_inc_vat":12.6,"valid_from":"2023-07-19T12:00:00Z","valid_to":"2023-07-19T12:30:00Z","payment_method":null},{"value_exc_vat":13.96,"value_inc_vat":14.658,"valid_from":"2023-07-19T11:30:00Z","valid_to":"2023-07-19T12:00:00Z","payment_method":null},{"value_exc_vat":14.0,"value_inc_vat":14.7,"valid_from":"2023-07-19T11:00:00Z","valid_to":"2023-07-19T11:30:00Z","payment_method":null},{"value_exc_vat":15.46,"value_inc_vat":16.233,"valid_from":"2023-07-19T10:30:00Z","valid_to":"2023-07-19T11:00:00Z","payment_method":null},{"value_exc_vat":16.0,"value_inc_vat":16.8,"valid_from":"2023-07-19T10:00:00Z","valid_to":"2023-07-19T10:30:00Z","payment_method":null},{"value_exc_vat":15.4,"value_inc_vat":16.17,"valid_from":"2023-07-19T09:30:00Z","valid_to":"2023-07-19T10:00:00Z","payment_method":null},{"value_exc_vat":16.32,"value_inc_vat":17.136,"valid_from":"2023-07-19T09:00:00Z","valid_to":"2023-07-19T09:30:00Z","payment_method":null},{"value_exc_vat":16.48,"value_inc_vat":17.304,"valid_from":"2023-07-19T08:30:00Z","valid_to":"2023-07-19T09:00:00Z","payment_method":null},{"value_exc_vat":17.02,"value_inc_vat":17.871,"valid_from":"2023-07-19T08:00:00Z","valid_to":"2023-07-19T08:30:00Z","payment_method":null},{"value_exc_vat":17.94,"value_inc_vat":18.837,"valid_from":"2023-07-19T07:30:00Z","valid_to":"2023-07-19T08:00:00Z","payment_method":null},{"value_exc_vat":17.76,"value_inc_vat":18.648,"valid_from":"2023-07-19T07:00:00Z","valid_to":"2023-07-19T07:30:00Z","payment_method":null},{"value_exc_vat":20.0,"value_inc_vat":21.0,"valid_from":"2023-07-19T06:30:00Z","valid_to":"2023-07-19T07:00:00Z","payment_method":null},{"value_exc_vat":15.6,"value_inc_vat":16.38,"valid_from":"2023-07-19T06:00:00Z","valid_to":"2023-07-19T06:30:00Z","payment_method":null},{"value_exc_vat":21.12,"value_inc_vat":22.176,"valid_from":"2023-07-19T05:30:00Z","valid_to":"2023-07-19T06:00:00Z","payment_method":null},{"value_exc_vat":16.0,"value_inc_vat":16.8,"valid_from":"2023-07-19T05:00:00Z","valid_to":"2023-07-19T05:30:00Z","payment_method":null},{"value_exc_vat":13.88,"value_inc_vat":14.574,"valid_from":"2023-07-19T04:30:00Z","valid_to":"2023-07-19T05:00:00Z","payment_method":null},{"value_exc_vat":13.11,"value_inc_vat":13.7655,"valid_from":"2023-07-19T04:00:00Z","valid_to":"2023-07-19T04:30:00Z","payment_method":null},{"value_exc_vat":13.42,"value_inc_vat":14.091,"valid_from":"2023-07-19T03:30:00Z","valid_to":"2023-07-19T04:00:00Z","payment_method":null},{"value_exc_vat":13.4,"value_inc_vat":14.07,"valid_from":"2023-07-19T03:00:00Z","valid_to":"2023-07-19T03:30:00Z","payment_method":null},{"value_exc_vat":13.42,"value_inc_vat":14.091,"valid_from":"2023-07-19T02:30:00Z","valid_to":"2023-07-19T03:00:00Z","payment_method":null},{"value_exc_vat":14.4,"value_inc_vat":15.12,"valid_from":"2023-07-19T02:00:00Z","valid_to":"2023-07-19T02:30:00Z","payment_method":null},{"value_exc_vat":12.24,"value_inc_vat":12.852,"valid_from":"2023-07-19T01:30:00Z","valid_to":"2023-07-19T02:00:00Z","payment_method":null},{"value_exc_vat":13.12,"value_inc_vat":13.776,"valid_from":"2023-07-19T01:00:00Z","valid_to":"2023-07-19T01:30:00Z","payment_method":null},{"value_exc_vat":13.54,"value_inc_vat":14.217,"valid_from":"2023-07-19T00:30:00Z","valid_to":"2023-07-19T01:00:00Z","payment_method":null},{"value_exc_vat":14.0,"value_inc_vat":14.7,"valid_from":"2023-07-19T00:00:00Z","valid_to":"2023-07-19T00:30:00Z","payment_method":null},{"value_exc_vat":13.96,"value_inc_vat":14.658,"valid_from":"2023-07-18T23:30:00Z","valid_to":"2023-07-19T00:00:00Z","payment_method":null},{"value_exc_vat":14.38,"value_inc_vat":15.099,"valid_from":"2023-07-18T23:00:00Z","valid_to":"2023-07-18T23:30:00Z","payment_method":null}]}
```

where the average rate should be reported as **16.548**. If we had the worse possible scenario where the last hour was negative (e.g. -9 on both), then I would expect this average to change at ~16:00 to **13.7175** which isn't a large margin of error in my opinion. If I started including data from the previous day, then this would undoubtly cause additional issues to be raised. But that's why I make the data available so you can make your own assumptions if the ones in the integration don't fit your needs (like Andy did).

> The value changes every 30 mins or so 

You said that the average rate changes every 30 minutes or so. Do you have examples of what the data changes to?

**Comment by d1964b** [comment]
OK , so I'm misreading this. I had created an Entity card to display the 'Current day average rate' attribute and had assumed that the graph automatically created by that card and visible when selecting the icon in the top right hard corner would be of the value the card was displaying. Sadly not. A fact I only realised after you confirmed the average rate was based on the date which forced me to look again at the card and past my own assumptions about it. The value does indeed seem to change a couple of times a day , at midnight and again at 1600 , approx., when the rest of todays prices come in. 
I'll accept 90% of the blame on this one , but 10% I am laying at HA's door for being so  bl__dy counter intuitive. Not everyone will agree with me but I'll die on that hill.
I'm curious how you draw a graph like the one above , perhaps someone could say.

**Comment by deltamelter** [comment]
> I'm curious how you draw a graph like the one above , perhaps someone could say.

Apex charts card, it takes the current_rate entity as an array of datetimes and values and displays a graph.  I also added conditional colouring, the average and the Now lines.   mini-graph-card is easier to setup, but also produces simpler graph features
